### ISSUE TITLE
nunit-agent.exe not terminating form time to time

### ISSUE THREAD

**Issue by michael77777777777** [nunit/nunit-console#1350]
**Error Description**
We have a gitlab-runner Windows service installed on our "Windows11 Pro 22H2" build machine. This service allows up to 4 parallel gitlab-pipelines. In each gitlab-pipeline we call _nunit3-console.exe_ 3.16.3 to execute our test assembly with nunit 3.13.3.

From time to time we observe that the _nunit-agent.exe_ process ist not terminated.
Using Microsofts "ProcessExplorer" tool to analyze the stuck _nunit-agent.exe_  process shows that its parent process is not existing anymore. Moreover, when switching to the "Threads" tab in ProcessExplorer, a Windows internal process lifecycle-check must be triggered as the stuck _nunit-agent.exe_  terminates immediately. 

In MsBuild file, we call _nunit3-console.exe_ with the following parameters_
`nunit3-console.exe $(testAssemblyList) --dispose-runners --skipnontestassemblies --framework=&quot;net-4.5&quot; --where=&quot;$(nunit_where)&quot; --encoding=utf-8 --labels=all --result=TestResult.xml  --result=junit-results.xml;transform=nunit3-junit.xslt`

**Remarks:**
* the bug occurs rarely and we did not find any possibility to reproduce it. We observed a higher occurrence of the bug when gitlab runners were executed parallel
* there is only one test assembly, so executing with flag "inprocess" has no effects
* our tests do not use threads

**Logging:**
Gitlab console logfile shows that all tests have been executed successfully before process get stuck

> LAST UNIT TEST IN THE LOGFILE
>
> Run Settings
>       RequestedRuntimeFramework: net-4.5
>       DisposeRunners: True
>       SkipNonTestAssemblies: True
>       WorkDirectory: C:\Tools\gitlab-runner\builds\moKKEX1Y\0\******
>       ImageRuntimeVersion: 4.0.30319
>       ImageRequiresX86: False
>       ImageRequiresDefaultAppDomainAssemblyResolver: False
>       TargetRuntimeFramework: net-4.5
>       NumberOfTestWorkers: 20
>
> Test Run Summary
>     Overall result: Passed
>     Test Count: 293, Passed: 293, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
>     Start time: 2023-07-05 07:08:52Z
>       End time: 2023-07-05 07:23:51Z
>       Duration: 898.792 seconds
>   
>   Results (nunit3) saved as TestResult.xml
>   Results (user) saved as junit-results.xml
>   
>   NUnit.Engine.NUnitEngineUnloadException : Agent Process was terminated successfully after error.
>     ----> NUnit.Engine.NUnitEngineUnloadException : Unable to unload application domain: unload thread timed out after 30 seconds.
>   Application domain name: domain-
>   Application domain BaseDirectory: C:\Tools\gitlab-runner\builds\moKKEX1Y\0\*******

**Assumption:**
Due to the behavior of ProcessExplorer this seems to be a threading issue. Maybe due to open TCP-connections?

**Machine configuration:**
* windows 11 pro 22h2 build server
* gitlab-runner 15.9.1 windows service with up to 4 parallel gitlab-pipelines
* nunit3-console.exe 3.16.3 from NuGet
* nunit 3.13.3 from NuGet

Any help is highly appreciated. Thank you in advance!

**Comment by michael77777777777** [comment]
Is there really nobody who can help with this issue?  Still facing this very anoying problem.

**Comment by michael77777777777** [comment]
Thanks for the reply @OsirisTerje . I understand your situation. I try to figure out more background information the next time it occurrs. Maybe something new catches my eyes...

**Comment by CharliePoole** [comment]
I'm going to close this for lack of a repro - understanding that a repro is virtually impossible here.

In order to avoid the runner itself hanging, if it loses contact with the agent or is unable to convince it to shut down, the runner just reports and ignores the situation. I think what is really needed is an ability to see what has happened within the agent that is hanging. If anyone has specific ideas about this, feel free to create a new issue..

**Comment by michael77777777777** [comment]
One remark from my side regarding this issue.

I refactored all test projects of my solution and made sure each test runs independently from the other ones. I also ensured that each test has its own ressources and they are tore down after test completion ( i.e. "I" and "R" in the FIRST pattern)

After this refactoring, the issue did not happen anymore. Unfortunately, due to the massive refactoring I cannot tell the actual cause. But it is a strategy that may help other people.
